### PR TITLE
Bulk I/O - v3

### PR DESCRIPTION
**Pull request by bbockelm**
This is the second attempt for merging the bulk IO APIs (hidden inside the ROOT::Internal namespace); original version was in #943; this is a large rebase and update (particularly, removing the buffer sharing between the ROOT and the user). The initial goal here is to get another round of @pcanal attention and feedback.

The bulk IO APIs aim to provide as fast access as possible in the case where the user is using relatively simple datatypes and the objects can be deserialized "non-destructively" (i.e., in-place in the existing memory buffers). For example - primitives and lists of primitives are supported; anything needing references are not. Despite the fact many of ROOT's very useful deserialization features are not supported, the limited feature set covers a range of analysis use cases. In particular, the current version of CMS's proposed NanoAOD format can be read solely via bulk APIs.

Where the bulk IO APIs are applicable, we see an order-magnitude improvement in event rates for some use cases (including CMS NanoAOD). The speedup becomes more noticeable as the data read from the event is smaller (meaning the traditional ROOT IO overheads are larger). Somewhere in the neighborhood of 5KB objects read per event is the point where bulk IO APIs should be noticeably faster.

This PR provides three interfaces:

   * A raw, low-level interface. The user can invoke the bulk IO APIs directly and get a buffer of primitive types, serialized or deserialized. It is not envisioned this is used by users, but by API writers.
   * A TTreeReader-like interface, TTreeReaderFast. This is meant to be used by users; it utilizes the low-level interface to retrieve the serialized buffer and templated code will inline the deserializing as part of iterating through the tree. The speed here is gained by the fact only a single pass in memory is needed to deserialize and apply user code, rather than two.
      * The intent is that this interface can be extended in the future and used by TDataFrame, provided we can make TDataFrame sufficiently fast.
   * A Python-based numpy export interface. This exposes the (possibly serialized) memory directly as numpy arrays, allowing for efficient access to ROOT-serialized objects from the Python language.
      * The numpy interface is low-level for Python; we intend to further explore the ideal high-level interface through projects such as https://github.com/diana-hep/plur

(With many thanks to @jpivarski for driving this work home with his work on a python interface!)

### CODE REVIEW

**Comment by bbockelm** [comment]
@phsft-bot build

File: bindings/pyroot/src/NumpyInterface.cxx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    import_array();
+    return 0;
+  }
+#else
+  void InitializeNumpy() {
+    import_array();
+  }
+#endif
+
+  PyObject* FillNumpyArray(PyObject* self, PyObject* args) {
```
Comment (etejedor) [inline bindings/pyroot/src/NumpyInterface.cxx]: @bbockelm @jpivarski I wonder if we could integrate this new pythonization with the current API we have to read TTrees into numpy arrays, in particular I am talking about [AsMatrix](https://github.com/root-project/root/blob/master/bindings/pyroot/ROOT.py#L263).

`FillNumpyArray` reads a given branch with bulk I/O, while `AsMatrix` can also read a single branch of a TTree into a numpy array with regular I/O. Maybe we could think of an implementation of `AsMatrix` that supersedes `FillNumpyArray` and uses bulk I/O.

**Comment by jpivarski** [comment]
@etejedor That sounds like a good idea. I've said before that this diff is so old that the Numpy parts can be cleanly removed and reimplemented on top of new developments from the past year. That should include TTree::AsMatrix.

I was also thinking of some jagged and object types, which are considered "destructive" storage by BulkIO, yet are not too hard to reconstruct. The code I've written to interpret these raw buffers as structured data has been separated out of uproot so that it can be used independently*. I'll present that on Monday at the ROOT Workshop and we could talk about it then.

* For instance, PyROOT could "optionally" depend on them, only performing a Python import inside a method call. The two libraries I'm taking about are awkward-array (for jaggedness and similarly difficult array structures) and uproot-methods (for Pythonic array-at-a-time operations).

**Comment by bbockelm** [comment]
@etejedor - given the comment from @jpivarski -- what do you think the best way forward is?  Delay the numpy piece until later?

I don't have a strong opinion here, as long as we can continue to push things in the right direction.

**Comment by jpivarski** [comment]
@etejedor @bbocjelm I think a simple way to do it is to remove the differences in this pull request that touch PyROOT— my contributions. Then this pull request can concentrate on only the TBranch/TBasket/TBufferFile modifications in C++ and we can later add code to AsMatrix to make use of it.

A year ago, AsMatrix didn't exist, so the Numpy interface was a one-off. Now that there's AsMatrix, there's a natural place to put it. Nothing in git is every lost— Enric can see how I implemented it and copy that.

**Comment by bbockelm** [comment]
@etejedor @jpivarski - done!

I kept the git history per request.

File: tree/tree/inc/TBranch.h [hunk 2]
```diff
@@ -115,6 +122,7 @@ class TBranch : public TNamed , public TAttFill {
    TBuffer    *fEntryBuffer;      ///<! Buffer used to directly pass the content without streaming
    TBuffer    *fTransientBuffer;  ///<! Pointer to the current transient buffer.
    TList      *fBrowsables;       ///<! List of TVirtualBranchBrowsables used for Browse()
+   std::unique_ptr<ROOT::Experimental::Internal::TBulkBranchRead> fBulk; ///<! Helper for performing bulk IO
```
Comment (pcanal) [inline tree/tree/inc/TBranch.h]: To reduce the code length, it might be worth have a local using/alias statement for TBulkBranchRead

File: tree/tree/inc/TLeaf.h [hunk 3]
```diff
@@ -59,6 +59,13 @@ class TLeaf : public TNamed {
       kNewValue = BIT(12)         ///< Set if we own the value buffer and so must delete it ourselves.
    };
 
+   enum DeserializeType{
```
Comment (pcanal) [inline tree/tree/inc/TLeaf.h]: missing white space before {

File: tree/tree/inc/TLeaf.h [hunk 4]
```diff
@@ -59,6 +59,13 @@ class TLeaf : public TNamed {
...
+      kInvalid = 0,      // Invalid deserialization information.
+      kDestructive = 0,  // Deserialization of this Leaf requires a separate output buffer.
```
Comment (pcanal) [inline tree/tree/inc/TLeaf.h]: Why is Invalid and Destructive the same value?

File: tree/tree/inc/TLeafI.h [hunk 5]
```diff
@@ -49,6 +50,8 @@ class TLeafI : public TLeaf {
    virtual void    PrintValue(Int_t i=0) const;
    virtual void    ReadBasket(TBuffer &b);
    virtual void    ReadBasketExport(TBuffer &b, TClonesArray *list, Int_t n);
+   virtual bool    ReadBasketFast(TBuffer&, Long64_t);
+   virtual bool    ReadBasketSerialized(TBuffer&, Long64_t) {return GetDeserializeType() == kInPlace; }
```
Comment (pcanal) [inline tree/tree/inc/TLeafI.h]: The semantic of this routine are unclear (i.e. need at least a doc string).   Why are the arguments never used?

File: tree/tree/src/TBasket.cxx [hunk 6]
```diff
@@ -699,6 +699,26 @@ Int_t TBasket::ReadBasketBytes(Long64_t pos, TFile *file)
...
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Disown all references to the internal buffer - some other object likely now
+/// owns it.
+///
+/// This TBasket is now useless and invalid until it is told to adopt a buffer.
+void TBasket::DisownBuffer()
+{
+   fBufferRef = NULL;
```
Comment (pcanal) [inline tree/tree/src/TBasket.cxx]: What if this is called when the TBaskets owns the fBufferRef?

File: tree/tree/src/TBranch.cxx [hunk 7]
```diff
@@ -115,6 +118,7 @@ TBranch::TBranch()
 , fEntryBuffer(0)
 , fTransientBuffer(0)
 , fBrowsables(0)
+, fBulk(new ROOT::Experimental::Internal::TBulkBranchRead(*this))
```
Comment (pcanal) [inline tree/tree/src/TBranch.cxx]: humm ... this means that this is allocated for every branch no matter what.  Is that really what we want? (i.e. price paid when not using bulk I/O).

File: tree/tree/src/TBranch.cxx [hunk 8]
```diff
@@ -1370,9 +1366,204 @@ Int_t TBranch::GetEntry(Long64_t entry, Int_t getall)
                GetBasket(i);
             }
          }
+         // Getting the next basket might reset the current one and
+         // cause a reset of the first / next basket entries back to -1.
+         fFirstBasketEntry = first;
+         fNextBasketEntry = updatedNext;
+      }
+      if (user_buffer) {
```
Comment (pcanal) [inline tree/tree/src/TBranch.cxx]: Can we add a comment explaining why if we have a user_basket, basket is not the 'currentBasket'?

File: tree/tree/src/TBranch.cxx [hunk 9]
```diff
@@ -1370,9 +1366,204 @@ Int_t TBranch::GetEntry(Long64_t entry, Int_t getall)
...
+   if (R__unlikely(entry != first)) {
+       //printf("Failed to read from full cluster; first entry is %ld; requested entry is %ld.\n", first, entry);
+       return -1;
+   }
+
+   basket->PrepareBasket(entry);
+   TBuffer* buf = basket->GetBufferRef();
+
+   // Test for very old ROOT files.
+   if (R__unlikely(!buf)) {Error("GetEntriesFast", "Failed to get a new buffer.\n"); return -1;}
```
Comment (pcanal) [inline tree/tree/src/TBranch.cxx]: split on top the next line if more than one statement in if.

File: tree/tree/src/TBranch.cxx [hunk 10]
```diff
@@ -1370,9 +1366,204 @@ Int_t TBranch::GetEntry(Long64_t entry, Int_t getall)
...
+   fBaskets[fReadBasket] = nullptr;
+   fExtraBasket = basket;
+   basket->DisownBuffer();
+
+   return N;
+}
+
+// TODO: Template this and the call above; only difference is the TLeaf function (ReadBasketFast vs
+// ReadBasketSerialized
+Int_t
```
Comment (pcanal) [inline tree/tree/src/TBranch.cxx]: No newline after return type.

File: tree/tree/src/TBranch.cxx [hunk 11]
```diff
@@ -1370,9 +1366,204 @@ Int_t TBranch::GetEntry(Long64_t entry, Int_t getall)
...
+TBranch::GetEntriesSerialized(Long64_t entry, TBuffer &user_buf, TBuffer *count_buf)
+{
+   // TODO: eventually support multiple leaves.
+   if (R__unlikely(fNleaves != 1)) {return -1;}
```
Comment (pcanal) [inline tree/tree/src/TBranch.cxx]: space after { and before }

File: tree/tree/src/TLeafElement.cxx [hunk 12]
```diff
@@ -62,6 +63,68 @@ TLeafElement::~TLeafElement()
...
+      return kInPlace;
+   }
+
+   fDeserializeTypeCache.store(kDestructive, std::memory_order_relaxed);
+   return kDestructive;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Deserialize N events from an input buffer.
+bool
```
Comment (pcanal) [inline tree/tree/src/TLeafElement.cxx]: no new line after return type.  use Bool_t.

File: tree/tree/test/BulkApi.cxx [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    ASSERT_TRUE(tree);
+
+    TBranch *branchF = tree->GetBranch("myFloat");
+    ASSERT_TRUE(branchF);
+
+    Int_t events = fEventCount;
+    float idx_f = 1;
+    Long64_t evt_idx = 0;
+    while (events) {
+        auto count = branchF->GetBulkRead().GetEntriesSerialized(evt_idx, branchbuf);
```
Comment (pcanal) [inline tree/tree/test/BulkApi.cxx]: 3 characters indent.

File: tree/tree/inc/TLeaf.h (hunk 4 above)
Comment (bbockelm) [reply tree/tree/inc/TLeaf.h]: looks like a rebase issue - should be distinct.

File: tree/tree/src/TBasket.cxx (hunk 6 above)
Comment (bbockelm) [reply tree/tree/src/TBasket.cxx]: Really, this is part of the contractual relationship between TBranch and TBasket.  Will make this into a private method.

**Comment by bbockelm** [comment]
@pcanal - this one is ready for another round of review.  I think Zhe and I hit all your outstanding comments.

File: tree/tree/inc/TBasket.h [hunk 14]
```diff
@@ -108,10 +113,8 @@ class TBasket : public TKey {
    TBasket(const char *name, const char *title, TBranch *branch);
    virtual ~TBasket();
 
-           void    AdoptBuffer(TBuffer *user_buffer);
    virtual void    AdjustSize(Int_t newsize);
    virtual void    DeleteEntryOffset();
-           void    DisownBuffer();
```
Comment (amadio) [inline tree/tree/inc/TBasket.h]: This is a backward-incompatible change in interface. I'm not sure it's a good idea. For example, why did you need to declare `TBranch` as a friend after the change in interface?

Comment (bbockelm) [reply tree/tree/inc/TBasket.h]: Actually, this method was introduced in this branch - it doesn't exist currently.

The change was motivated by a comment from @pcanal - the `Adopt` / `Disown` methods are a specific protocol between `TBranch` and its child `TBasket`.  They would be difficult to use outside the `TBranch`.

To me, it made more sense for `TBranch` / `TBasket` to be friends rather than expose memory ownership concerns to anyone invoking ROOT code.

Would welcome feedback on the approach!

File: tree/tree/inc/TBranch.h [hunk 15]
```diff
@@ -134,14 +143,19 @@ class TBranch : public TNamed , public TAttFill {
...
    TBasket *GetFreshCluster();
    Int_t    WriteBasket(TBasket* basket, Int_t where) { return WriteBasketImpl(basket, where, nullptr); }
 
    TString  GetRealFileName() const;
 
 private:
-   Int_t FillEntryBuffer(TBasket* basket,TBuffer* buf, Int_t& lnew);
+   Int_t    GetBasketAndFirst(TBasket*& basket, Long64_t& first, TBuffer* user_buffer);
+   TBasket *GetBasketImpl(Int_t basket, TBuffer* user_buffer);
+   Int_t    GetEntriesFast(Long64_t, TBuffer&);
```
[the remark this answers is not in the document]
Comment (zzxuanyuan) [reply tree/tree/inc/TBranch.h]: Do you have any suggestions about the function names?

[the remark this answers is not in the document]
Comment (pcanal) [reply tree/tree/inc/TBranch.h]: GetBulkEntries?

[the remark this answers is not in the document]
Comment (zzxuanyuan) [reply tree/tree/inc/TBranch.h]: Sounds good to me.

File: tree/tree/src/TLeafD.cxx [hunk 16]
```diff
@@ -129,6 +130,22 @@ void TLeafD::ReadBasket(TBuffer &b)
    }
 }
 
+// Deserialize N events from an input buffer.
+bool TLeafD::ReadBasketFast(TBuffer &input_buf, Long64_t N) {
+  if (R__unlikely(fLeafCount)) { return false; }
+
+   Double_t *buf __attribute__((aligned(8)));
+   buf = reinterpret_cast<Double_t*>(input_buf.GetCurrent());
```
[the remark this answers is not in the document]
Comment (pcanal) [reply tree/tree/src/TLeafD.cxx]: There is a #ifdef R__BYTESWAP, don't we need an #else clause?

**Comment by pcanal** [comment]
This PR is continued at https://github.com/root-project/root/pull/3612
